README.md
# rpi-displays
A python library to put text to several displays which are connected with a raspberry


CHANGELOG.rst
Changelog
=========

## 0.1.2 - 2016-09-11

* use backlight var from init, not global


## 0.1.1 - 2016-09-11

* Turn display on in initial method to avoid blinking cursor


## 0.1.0 - 2016-09-11

* Initial release


rpi_displays/sainsmart/displays.py
# -*- coding: utf-8 -*-
from time import sleep

from ..channels import I2C
from .constants import (
    LCD_2LINE, LCD_4BITMODE, LCD_BACKLIGHT, LCD_CLEARDISPLAY, LCD_DISPLAYCONTROL, LCD_DISPLAYON,
    LCD_ENTRYLEFT, LCD_ENTRYMODESET, LCD_FUNCTIONSET, LCD_NOBACKLIGHT, LCD_RETURNHOME, En,
    LCD_5x8DOTS, Rs)


class LCD2004:
    """
    This is the driver for the SainSmart LCD 2004 models.
    It can be used for the 16x2 and 20x4
    """

    def __init__(
            self, address=0x27, lines=LCD_2LINE, dots=LCD_5x8DOTS, bitmode=LCD_4BITMODE,
            backlight=LCD_BACKLIGHT):
        self.device = I2C(address)
        self.backlight = backlight

        self.set(0x03)
        self.set(0x03)
        self.set(0x03)
        self.set(0x02)

        self.set(LCD_FUNCTIONSET | lines | dots | bitmode)
        self.set(LCD_DISPLAYCONTROL | LCD_DISPLAYON)
        self.set(LCD_ENTRYMODESET | LCD_ENTRYLEFT)
        sleep(0.2)

    # clocks EN to latch command
    def strobe(self, data):
        self.device.write_cmd(data | En | self.backlight)
        sleep(.0005)
        self.device.write_cmd(((data & ~En) | self.backlight))
        sleep(.0001)

    def set_four_bits(self, data):
        self.device.write_cmd(data | self.backlight)
        self.strobe(data)

    def set(self, cmd, mode=0):
        self.set_four_bits(mode | (cmd & 0xF0))
        self.set_four_bits(mode | ((cmd << 4) & 0xF0))

    def set_char(self, charvalue, mode=1):
        self.set_four_bits(mode | (charvalue & 0xF0))
        self.set_four_bits(mode | ((charvalue << 4) & 0xF0))

    # custom chars (0 - 7)
    def set_custom_chars(self, fontdata):
        self.set(0x40)
        for char in fontdata:
            for line in char:
                self.set_char(line)

    def display_string(self, string, line):
        if line == 1:
            self.set(0x80)
        if line == 2:
            self.set(0xC0)
        if line == 3:
            self.set(0x94)
        if line == 4:
            self.set(0xD4)

        for char in string:
            self.set(ord(char), Rs)

    def display_string_position(self, string, line, pos):
        if line == 1:
            pos_new = pos
        elif line == 2:
            pos_new = 0x40 + pos
        elif line == 3:
            pos_new = 0x14 + pos
        elif line == 4:
            pos_new = 0x54 + pos

        self.set(0x80 + pos_new)

        for char in string:
            self.set(ord(char), Rs)

    def switch_backlight(self, state):
        if state == 1:
            self.device.write_cmd(LCD_BACKLIGHT)
        elif state == 0:
            self.device.write_cmd(LCD_NOBACKLIGHT)

    def clear(self):
        self.set(LCD_CLEARDISPLAY)
        self.set(LCD_RETURNHOME)


rpi_displays/sainsmart/constants.py
# -*- coding: utf-8 -*-
# commands
LCD_CLEARDISPLAY = 0x01
LCD_RETURNHOME = 0x02
LCD_ENTRYMODESET = 0x04
LCD_DISPLAYCONTROL = 0x08
LCD_CURSORSHIFT = 0x10
LCD_FUNCTIONSET = 0x20
LCD_SETCGRAMADDR = 0x40
LCD_SETDDRAMADDR = 0x80

# flags for display entry mode
LCD_ENTRYRIGHT = 0x00
LCD_ENTRYLEFT = 0x02
LCD_ENTRYSHIFTINCREMENT = 0x01
LCD_ENTRYSHIFTDECREMENT = 0x00

# flags for display on/off control
LCD_DISPLAYON = 0x04
LCD_DISPLAYOFF = 0x00
LCD_CURSORON = 0x02
LCD_CURSOROFF = 0x00
LCD_BLINKON = 0x01
LCD_BLINKOFF = 0x00

# flags for display/cursor shift
LCD_DISPLAYMOVE = 0x08
LCD_CURSORMOVE = 0x00
LCD_MOVERIGHT = 0x04
LCD_MOVELEFT = 0x00

# flags for function set
LCD_8BITMODE = 0x10
LCD_4BITMODE = 0x00
LCD_2LINE = 0x08
LCD_1LINE = 0x00
LCD_5x10DOTS = 0x04
LCD_5x8DOTS = 0x00

# flags for backlight control
LCD_BACKLIGHT = 0x08
LCD_NOBACKLIGHT = 0x00

En = 0b00000100  # Enable bit
Rw = 0b00000010  # Read/Write bit
Rs = 0b00000001  # Register select bit


rpi_displays/examples/sainsmart/lcd_2004.py
from time import sleep

from rpi_displays.sainsmart.displays import LCD2004


# a custom icon, consisting of 6 individual characters
# 3 chars in the first row and 3 chars in the second row
fontdata1 = [
    [0x00, 0x00, 0x03, 0x04, 0x08, 0x19, 0x11, 0x10],
    [0x00, 0x1F, 0x00, 0x00, 0x00, 0x11, 0x11, 0x00],
    [0x00, 0x00, 0x18, 0x04, 0x02, 0x13, 0x11, 0x01],
    [0x12, 0x13, 0x1b, 0x09, 0x04, 0x03, 0x00, 0x00],
    [0x00, 0x11, 0x1f, 0x1f, 0x0e, 0x00, 0x1F, 0x00],
    [0x09, 0x19, 0x1b, 0x12, 0x04, 0x18, 0x00, 0x00],
    [0x1f, 0x0, 0x4, 0xe, 0x0, 0x1f, 0x1f, 0x1f],
]

fontdata2 = [
    [0x1, 0x3, 0x7, 0xf, 0xf, 0x7, 0x3, 0x1],
    [0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10, 0x10],
    [0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18, 0x18],
    [0x1c, 0x1c, 0x1c, 0x1c, 0x1c, 0x1c, 0x1c, 0x1c],
    [0x1e, 0x1e, 0x1e, 0x1e, 0x1e, 0x1e, 0x1e, 0x1e],
    [0x0, 0x1, 0x3, 0x7, 0xf, 0x1f, 0x1f, 0x1f],
]

lcd = LCD2004()
lcd.display_string('RPi I2C test', 1)
lcd.display_string(' Custom chars', 2)

sleep(2)

lcd.clear()

# Set logo chars (fontdata1)
lcd.set_custom_chars(fontdata1)

# Write first three chars to row 1 directly
lcd.set(0x80)
lcd.set_char(0)
lcd.set_char(1)
lcd.set_char(2)

# Write next three chars to row 2 directly
lcd.set(0xC0)
lcd.set_char(3)
lcd.set_char(4)
lcd.set_char(5)
sleep(2)

lcd.clear()

lcd.display_string_position('Testing', 1, 1)
sleep(1)

lcd.display_string_position('Testing', 2, 3)
sleep(1)
lcd.clear()

# Set logo chars from the second set
lcd.set_custom_chars(fontdata2)

block = chr(255)  # block character, built-in

# display two blocks in columns 5 and 6 (i.e. AFTER pos. 4) in row 1
# first draw two blocks on 5th column (cols 5 and 6), starts from 0
lcd.display_string_position(block * 2, 1, 4)

# chars starting from col. 7 (pos. 6)
pos = 6
lcd.display_string_position(chr(1), 1, 6)
sleep(0.2)

lcd.display_string_position(chr(2), 1, pos)
sleep(0.2)

lcd.display_string_position(chr(3), 1, pos)
sleep(0.2)

lcd.display_string_position(chr(4), 1, pos)
sleep(0.2)

lcd.display_string_position(block, 1, pos)
sleep(0.2)


# another one, same as above, 1 char-space to the right
pos = pos + 1  # increase column by one

lcd.display_string_position(chr(1), 1, pos)
sleep(0.2)

lcd.display_string_position(chr(2), 1, pos)
sleep(0.2)

lcd.display_string_position(chr(3), 1, pos)
sleep(0.2)

lcd.display_string_position(chr(4), 1, pos)
sleep(0.2)

lcd.display_string_position(block, 1, pos)
sleep(0.2)


# set first set of custom chars - smiley
lcd.set_custom_chars(fontdata1)

lcd.display_string_position(chr(0), 1, 9)
lcd.display_string_position(chr(1), 1, 10)
lcd.display_string_position(chr(2), 1, 11)
lcd.display_string_position(chr(3), 2, 9)
lcd.display_string_position(chr(4), 2, 10)
lcd.display_string_position(chr(5), 2, 11)
sleep(2)

lcd.clear()
sleep(1)

lcd.switch_backlight(0)
